sdv/single_table/utils.py

```diff
@@ -27,24 +31,28 @@ def detect_discrete_columns(metadata, data):
             A list of discrete columns to be used with some of ``sdv`` synthesizers.
     """
     discrete_columns = []
-
     for column in data.columns:
-        if column in metadata.columns:
-            if metadata.columns[column]['sdtype'] not in ['numerical', 'datetime']:
-                discrete_columns.append(column)
+        # Numerical and datetime columns never get preprocessed into categorical ones
+        if column in metadata.columns and \
+                metadata.columns[column]['sdtype'] in ['numerical', 'datetime']:
+            continue
 
-        else:
-            column_data = data[column].dropna()
-            if set(column_data.unique()) == {0.0, 1.0}:
-                column_data = column_data.astype(bool)
+        column_data = data[column].dropna()
+
+        # Ignore columns with only nans and empty datasets
+        if column_data.empty:
+            continue
 
-            try:
-                dtype = column_data.infer_objects().dtype.kind
-                if dtype in ['O', 'b']:
-                    discrete_columns.append(column)
+        # Non-integer floats cannot be categorical
+        try:
+            column_data = column_data.astype('float')
+            if list(column_data) != list(column_data.round()):
+                continue
+        except BaseException:
+            pass
 
-            except Exception:
-                discrete_columns.append(column)
+        # Everything else is presumed categorical
+        discrete_columns.append(column)
```

Comment:
I think we also wanted to make sure the number of unique values is less than a threshold percentage of the total number of values in the column. I think we were going to use 10 or 20% as the threshold. Right @npatki